### PR TITLE
[WIP] work on returnTicker

### PR DESCRIPTION
**Pull request by JonathonDunford**
We will need another database table for ticker information.

Proposed information to store:
```
ticker = {
        'tokenAddr': ,
        'quoteVolume': ,
        'baseVolume': ,
        'last': ,
        'bid': ,
        'ask': ,   
    }
```

bid and ask can come from `orders`
last can come from `trades`

Would volume have to be determined by adding all of the orders in our orderbook?

### CODE REVIEW

**Comment by freeatnet** [comment]
How and when do we update `bid` and `ask`? Off the top of my head, they should at least change when new orders are created (and the price is higher / lower than previous one) and when there is no more volume available at the price reflected. How do we fetch that?

**Comment by JonathonDunford** [comment]
We haven't finished the code for #8 but that could be where we change the bid/ask based on volume changes.

OR

We could have something that scans our own database on a regular frequency, compares it to ticker data, then changes ticker data where it sees fit.

We can invalidate orders separately and the change would be reflected there.

File: app/services/websocket_server.py [hunk 1]
```diff
@@ -231,6 +231,36 @@ def format_order(record):
 
     await sio.emit('market', response, room=sid)
 
+#return ticker information, if no token specified show all
+@sio.on('returnTicker')
```
Comment (freeatnet) [inline app/services/websocket_server.py]: There isn't a separate `returnTicker` event: instead, `returnTicker` is a key on a `market` [event](https://github.com/etherdelta/etherdelta.github.io/blob/master/docs/API.md#getmarket--token-address-user-address-), which is emitted to a specific user in response to a `getMarket` event by that user.

File: app/services/websocket_server.py [hunk 2]
```diff
@@ -202,24 +251,46 @@ def format_order(record):
...
     trades = []
     my_trades = []
     my_funds = []
+    tickers = {}
+    
+    #if token is passed in
     if token:
+        
+        #get ticker for passed in token
+        tickers = await get_tickers(token)
```
Comment (freeatnet) [inline app/services/websocket_server.py]: We should return all tickers regardless.

File: app/services/websocket_server.py [hunk 3]
```diff
@@ -22,6 +23,54 @@
 def connect(sid, environ):
     print("connect ", sid)
 
+#get returnTicker, grabs data from tickers table
+async def get_tickers(token_hexstr):
+
+    #init
```
Comment (freeatnet) [inline app/services/websocket_server.py]: Nitpick: according to [PEP8](https://www.python.org/dev/peps/pep-0008/#inline-comments), the standard Python code style guide, "They should start with a # and a single space."

File: app/services/websocket_server.py [hunk 4]
```diff
@@ -22,6 +23,54 @@
...
+    where = ''
```
Comment (freeatnet) [inline app/services/websocket_server.py]: `placeholder_args` should also be defined here.

File: app/services/websocket_server.py [hunk 5]
```diff
@@ -22,6 +23,54 @@
...
+     modified: 2018-01-15T15:53:00},
+"""
+def format_ticker(ticker):
+    contract = ERC20Token(ticker["token_address"])
+    return { "{}_{}".format("ETH", Web3.toHex(ticker["token_address"]) : {
+        "tokenAddr": Web3.toHex(ticker["token_address"]),
+        "quoteVolume": str(contract.denormalize_value(ticker["quote_volume"])),
+        "baseVolume": str(contract.denormalize_value(ticker["base_volume"])),
+        "last": str(contract.denormalize_value(ticker["last"])),
+        "percentChange": str(contract.denormalize_value(ticker["percent_change"])),
```
Comment (freeatnet) [inline app/services/websocket_server.py:69]: Not a 100% sure what the values are going to look like in raw format, we should double-check these conversions once we have them.

File: app/services/websocket_server.py [hunk 6]
```diff
@@ -22,6 +23,54 @@
...
+        "bid": str(contract.denormalize_value(ticker["bid"])),
+        "ask": str(contract.denormalize_value(ticker["ask"])),
+        "modified": datetime.now().isoformat()
```
Comment (freeatnet) [inline app/services/websocket_server.py]: What is the modified value supposed to reflect?

Comment (JonathonDunford) [reply app/services/websocket_server.py]: When the last time that ticker was updated

File: app/services/websocket_server.py (hunk 2 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: If you pass in a specific token, it shouldn't transmit unnecessary data?

Comment (freeatnet) [reply app/services/websocket_server.py]: This data fills out the values in the bottom-left ticker though, so it needs to be for all tokens.

File: app/services/websocket_server.py (hunk 6 above)
Comment (freeatnet) [reply app/services/websocket_server.py]: Sounds like it should come from the DB then. Is this a placeholder?

Comment (JonathonDunford) [reply app/services/websocket_server.py]: You're right. Whoops.

Comment (JonathonDunford) [reply app/services/websocket_server.py]: Fixed.

File: app/services/websocket_server.py (hunk 4 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: fixed.

File: app/services/websocket_server.py (hunk 3 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: Fixed.

File: app/services/websocket_server.py (hunk 2 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: Referenced here: https://github.com/forkdelta/backend-replacement/issues/21

Fixed now.

File: app/services/websocket_server.py [hunk 7]
```diff
@@ -22,6 +23,56 @@
...
+    return { "{}_{}".format("ETH", Web3.toHex(ticker["token_address"]) : {
+        "tokenAddr": Web3.toHex(ticker["token_address"]),
+        "quoteVolume": str(contract.denormalize_value(ticker["quote_volume"])),
+        "baseVolume": str(contract.denormalize_value(ticker["base_volume"])),
+        "last": str(contract.denormalize_value(ticker["last"])),
+        "percentChange": str(contract.denormalize_value(ticker["percent_change"])),
+        "bid": str(contract.denormalize_value(ticker["bid"])),
+        "ask": str(contract.denormalize_value(ticker["ask"])),
+        "modified": ticker["modified"]
+        # "modified": datetime.now().isoformat()
```
Comment (freeatnet) [inline app/services/websocket_server.py]: Dead lines should be removed.

File: app/services/websocket_server.py [hunk 8]
```diff
@@ -202,24 +253,47 @@ def format_order(record):
...
     my_trades = []
     my_funds = []
+    tickers = {}
+    
+    # if token is passed in
     if token:
+        
+        # get ticker for passed in token
+        #tickers = await get_tickers(token)
+        tickers = await get_tickers()
```
Comment (freeatnet) [inline app/services/websocket_server.py]: We could probably extract that to above if we no longer differentiate by token.

File: app/services/websocket_server.py [hunk 9]
```diff
@@ -22,6 +23,56 @@
...
+        return await conn.fetch(
+            """
+            SELECT *
+            FROM tickers
+            WHERE {}
+            ORDER BY token_address DESC
+            """.format(where),
+            *placeholder_args)
+
+# format ticker information to camelcase from underscore, sanitize, add modified date
```
Comment (freeatnet) [inline app/services/websocket_server.py:49]: Oh, also, suggestion! Python [docstrings](https://www.python.org/dev/peps/pep-0257/) might be a better way to do this

File: app/services/websocket_server.py [hunk 10]
```diff
@@ -202,24 +253,47 @@ def format_order(record):
...
     trades = []
     my_trades = []
     my_funds = []
+    tickers = {}
+    
+    # if token is passed in
     if token:
+        
+        # get ticker for passed in token
+        #tickers = await get_tickers(token)
```
Comment (freeatnet) [inline app/services/websocket_server.py]: Also dead code

File: app/services/websocket_server.py [hunk 11]
```diff
@@ -202,24 +253,47 @@ def format_order(record):
...
 
+    # response vars
     trades = []
     my_trades = []
     my_funds = []
+    tickers = {}
+    
+    # if token is passed in
     if token:
+
```
Comment (freeatnet) [inline app/services/websocket_server.py]: Unnecessary new lines

File: app/services/websocket_server.py (hunk 8 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: Fixed.

File: app/services/websocket_server.py (hunk 7 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: Placeholder for when I write the code that updates the data. Will remove at that time.

File: app/services/websocket_server.py (hunk 10 above)
Comment (JonathonDunford) [reply app/services/websocket_server.py]: Fixed.

**Comment by JonathonDunford** [comment]
Now, we need something that actually populates this data.

I think the best method may be using an observer to scan our `orders` table and comparing actual data to the _old_ data stored in `tickers`.

Maybe there is a better way to only update it on certain events, but this would be a catch all and we could cancel/update orders separately without fear of breaking this system.

**Comment by JonathonDunford** [comment]
Convo copied here for posterity.

> **BigBrother - Today at 10:56 AM**
> Well, instead of relying on ED data to update our tickers, we would scan our own orders + trades table to get bid/ask/volume from orders and last/percent from trades(edited)
> 
> **freeatnet - Today at 10:57 AM**
> With you on that!
> 
> **BigBrother - Today at 10:58 AM**
> And since the rest of the backend handles updating and making sure that data is accurate
> regardless of ED connectability
> That should be the most reliable source of data?
> That's my thinking at least.
> 
> **freeatnet - Today at 11:02 AM**
> Absolutely agreed. To clarify: In my mind, there are two ways of doing this:
> 1. Relying on data we store, we periodically update ticker data from scratch by looking up the right orders and trades.
> 2. Relying on data we process, we track certain events and update ticker incrementally, based on the context of the event
> 
> First way is much, much simpler. We may have revisit it for purposes of real-time updating in the future, but it should work for now.

File: app/services/ticker_observer.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                    'percentChange': '',
+                }
+                
+    # TODO: Submit the ticker_updates to `tickers`
+    # TODO: Fill modified date with "modified": datetime.now().isoformat()
+    
+    return
+
+# Main function. Grabs orders and trades, then scans through orders and trades to update `tickers`
+async def update_tickers():
```
Comment (freeatnet) [inline app/services/ticker_observer.py:145]: Main function should be called `main()` and be called if the module is the main module of the python invocation, [like so.](https://github.com/forkdelta/backend-replacement/blob/master/app/services/etherdelta_observer.py#L161)

File: app/services/ticker_observer.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    
+    orders = await get_orders();
```
Comment (freeatnet) [inline app/services/ticker_observer.py:147]: No semicolons needed if you have just one expression / statement on a line.

File: app/services/ticker_observer.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            """
+
+# Maybe there is a fancy SQL statement here that will make this faster? 
+async def get_orders():
+
+    async with App().db.acquire_connection() as conn:
+        return await conn.fetch(
+            """
+            SELECT token_give, amount_give, token_get, amount_get, expires
+            FROM orders
```
Comment (freeatnet) [inline app/services/ticker_observer.py:46]: Pretty good SELECT statement for pulling all orders!

File: app/services/ticker_observer.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+from time import time
+import socketio
+from web3 import Web3
+from datetime import datetime
+
+sio = socketio.AsyncServer()
+app = web.Application()
+sio.attach(app)
+
+ZERO_ADDR = "0x0000000000000000000000000000000000000000"
```
Comment (freeatnet) [inline app/services/ticker_observer.py:25]: We should extract these into reusable constants. `app/constants.py`, perhaps?

File: app/services/ticker_observer.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+            # If ticker updates doesn't already have this token, add it to our object to update.
+            if not hasattr(ticker_updates, order["token_get"]):
+                
+                ticker_updates[order["token_get"]] = {
+                    'bid': this_orders_bid.
+                }
+                
+            else:
+                # If there isn't a current recorded bid OR current recorded bid is lower than the bid in this order.
+                if not has_attr(ticker_updates[order["token_get"]], 'bid') || ticker_updates[order["token_get"]]["bid"] < this_orders_bid:
```
Comment (freeatnet) [inline app/services/ticker_observer.py:85]: The logical or operator is `or`.

**Review by freeatnet — COMMENTED** [review]
Some minor things pointed out.

I'm also not sure about iterating over all orders to build a ticker. Perhaps, the ticker updater should know a list of token addresses which it should update; then, for each token, fetch fields with specific SQL queries. 

For example, a ticker bid value can be obtained by running something like
```sql
SELECT MAX(amount_give / amount_get)
FROM orders
WHERE token_get = :zero_addr
    AND token_give = :token_addr
    AND (state = 'OPEN'::orderstate AND expires > :current_block AND available_volume > 0)
```
This approach is a little harder on the DB, but it is faster than iterating over all orders and is somewhat simpler to implement than partial updates.

**Comment by JonathonDunford** [comment]
Moved to #40

**Pull request closed without merging**